### PR TITLE
`ITOFFOLIGate`

### PR DESCRIPTION
**Pull request by vtomole**

### CODE REVIEW

File: qiskit_superstaq/custom_gates.py [hunk 1]
```diff
@@ -183,6 +185,94 @@ def __str__(self) -> str:
         return f"ParallelGates({args})"
 
 
+class IITOFFOLIGate(ControlledGate):
+    def __init__(self, label: Optional[str] = None, ctrl_state: Optional[Union[str, int]] = None):
```
Comment (richrines1) [inline qiskit_superstaq/custom_gates.py:189]: given that this takes a `ctrl_state` argument it should just be `ITOFFOLI`/`ICCX`. We can define `IIToffoli`/`IICCX` as aliases to the instantiation w/ `ctrl_state="00"`

File: qiskit_superstaq/custom_gates.py [hunk 2]
```diff
@@ -183,6 +185,94 @@ def __str__(self) -> str:
...
+            [
+                [0, 0, 0, 0, 1j, 0, 0, 0],
+                [0, 1, 0, 0, 0, 0, 0, 0],
+                [0, 0, 1, 0, 0, 0, 0, 0],
+                [0, 0, 0, 1, 0, 0, 0, 0],
+                [1j, 0, 0, 0, 0, 0, 0, 0],
+                [0, 0, 0, 0, 0, 1, 0, 0],
+                [0, 0, 0, 0, 0, 0, 1, 0],
+                [0, 0, 0, 0, 0, 0, 0, 1],
+            ],
```
Comment (richrines1) [inline qiskit_superstaq/custom_gates.py:226]: this should take into account `ctrl_state`

File: qiskit_superstaq/custom_gates.py [hunk 3]
```diff
@@ -183,6 +185,94 @@ def __str__(self) -> str:
...
+            base_gate=XGate(),
+        )
+
+    def _define(self) -> None:
+        qc = qiskit.QuantumCircuit(3, name="iitoffoli")
+        csgate = SGate().control(1)
+        qc.x(0)
+        qc.x(1)
+        qc.ccx(0, 1, 2)
+        qc.append(csgate, [0, 1])
```
Comment (richrines1) [inline qiskit_superstaq/custom_gates.py:207]: can use
```suggestion
        qc.cp(np.pi / 2, 0, 1)
```

File: qiskit_superstaq/custom_gates.py [hunk 4]
```diff
@@ -183,6 +185,94 @@ def __str__(self) -> str:
...
+        qc.x(0)
+        qc.x(1)
```
Comment (richrines1) [inline qiskit_superstaq/custom_gates.py:209]: this should take into account `ctrl_state`, e.g. something like
```suggestion
        if self.ctrl_state & 2:
            qc.x(0)
        if self.ctrl_state & 1:
            qc.x(1)
        qc.ccx(0, 1, 2)
        qc.append(csgate, [0, 1])
        if self.ctrl_state & 2:
            qc.x(0)
        if self.ctrl_state & 1:
            qc.x(1)
```

File: qiskit_superstaq/custom_gates.py [hunk 5]
```diff
@@ -183,6 +185,94 @@ def __str__(self) -> str:
         return f"ParallelGates({args})"
 
 
+class IITOFFOLIGate(ControlledGate):
```
[the remark this answers is not in the document]
Comment (vtomole) [reply qiskit_superstaq/custom_gates.py]: Done.

File: qiskit_superstaq/custom_gates.py (hunk 1 above)
Comment (vtomole) [reply qiskit_superstaq/custom_gates.py]: Done.

File: qiskit_superstaq/custom_gates.py (hunk 2 above)
Comment (vtomole) [reply qiskit_superstaq/custom_gates.py]: Done.

File: qiskit_superstaq/custom_gates.py (hunk 3 above)
Comment (vtomole) [reply qiskit_superstaq/custom_gates.py]: Done.

File: qiskit_superstaq/custom_gates.py (hunk 4 above)
Comment (vtomole) [reply qiskit_superstaq/custom_gates.py]: Done.

**Comment by vtomole** [comment]
@richrines1 Now ready for a re-review.

File: qiskit_superstaq/custom_gates.py [hunk 6]
```diff
@@ -183,6 +183,84 @@ def __str__(self) -> str:
...
+        self, label: Optional[np.ndarray] = None, ctrl_state: Optional[Union[str, int]] = None
+    ) -> None:
+        super().__init__(
+            "iccx",
+            3,
+            [-np.pi],
+            num_ctrl_qubits=2,
+            label=label,
+            ctrl_state=ctrl_state,
+            base_gate=qiskit.circuit.library.RXGate(-np.pi),
```
Comment (richrines1) [inline qiskit_superstaq/custom_gates.py:197]: did we decide against defining an `IXGate` for this?

File: qiskit_superstaq/custom_gates_test.py [hunk 7]
```diff
@@ -110,10 +110,45 @@ def test_parallel_gates() -> None:
...
+
+
 def test_custom_resolver() -> None:
     gates = [
         qiskit_superstaq.AceCR("+-"),
         qiskit_superstaq.ZZSwapGate(1.23),
+        qiskit_superstaq.ICCXGate(),
+        qiskit_superstaq.ICCXGate(ctrl_state="00"),
+        qiskit_superstaq.ICCXGate(ctrl_state="01"),
+        qiskit_superstaq.ICCXGate(ctrl_state="10"),
```
Comment (richrines1) [inline qiskit_superstaq/custom_gates_test.py:151]: can we also add these to [serialization_test.py](https://github.com/SupertechLabs/qiskit-superstaq/blob/de5ffcda7d3b5456785d034fb50e313ecb3dd142/qiskit_superstaq/serialization_test.py#L46) to make sure `ctrl_state` is preserved after deserialization?

File: qiskit_superstaq/custom_gates.py (hunk 6 above)
Comment (vtomole) [reply qiskit_superstaq/custom_gates.py:197]: We did because we end up having to define an `IXdgGate` and we thought that was a bit much.

Comment (richrines1) [reply qiskit_superstaq/custom_gates.py:197]: got it. this should be fine, down the road we may want to come back to it

File: qiskit_superstaq/custom_gates_test.py (hunk 7 above)
Comment (vtomole) [reply qiskit_superstaq/custom_gates_test.py:151]: Done.

**Review by richrines1 — APPROVED** [review]
lgtm!

**Pull request merged**